File: android/java/v2/AmpliApp/app/src/main/java/com/example/ampliapp/SegmentDestinationPlugin.java
package com.example.ampliapp;

import com.amplitude.core.Amplitude;
import com.amplitude.core.events.BaseEvent;
import com.amplitude.core.events.GroupIdentifyEvent;
import com.amplitude.core.events.IdentifyEvent;
import com.amplitude.core.platform.DestinationPlugin;
import com.segment.analytics.Analytics;
import com.segment.analytics.Properties;
import com.segment.analytics.Traits;

import java.util.Map;

public class SegmentDestinationPlugin extends DestinationPlugin {
    android.content.Context context;
    Analytics analytics;
    String SEGMENT_API_KEY;
    public SegmentDestinationPlugin(android.content.Context appContext, String segmentAPIKey) {
        this.context = appContext;
        this.SEGMENT_API_KEY = segmentAPIKey;
    }
    @Override
     public void setup(Amplitude amplitude) {
        super.setup(amplitude);
        analytics = new Analytics.Builder(this.context, SEGMENT_API_KEY)
                .build();

        Analytics.setSingletonInstance(analytics);
    }

    @Override
    public BaseEvent track(BaseEvent event) {
        Properties properties = new Properties();
        for (Map.Entry<String,Object> entry : event.getEventProperties().entrySet()) {
            properties.putValue(entry.getKey(),entry.getValue());
        }
        analytics.track(event.eventType, properties);
        return event;
    }

    @Override
    public IdentifyEvent identify(IdentifyEvent identifyEvent) {
        Traits traits =  new Traits();
        for (Map.Entry<String,Object> entry : identifyEvent.getUserProperties().entrySet()) {
            traits.putValue(entry.getKey(), entry.getValue());
        }
        Analytics.with(this.context).identify(traits);

        return identifyEvent;
    }

    @Override
    public GroupIdentifyEvent groupIdentify(GroupIdentifyEvent groupIdentifyEvent) {
        Traits traits =  new Traits();
        String groupId = groupIdentifyEvent.getGroups().entrySet().iterator().next().getKey();
        for (Map.Entry<String,Object> entry : groupIdentifyEvent.getGroups().entrySet()) {
            traits.putValue(entry.getKey(),entry.getValue());
        }
        Analytics.with(context).group(groupId, traits);

        return groupIdentifyEvent;
    }
}

File: android/kotlin/v2/AmpliApp/app/src/main/java/com/example/ampliapp/SegmentDestinationPlugin.kt
//
// The example of destination plugin
// This destination plugin will also send event to Segment
//
package com.example.ampliapp

import android.content.Context
import com.amplitude.core.events.GroupIdentifyEvent
import com.amplitude.core.events.IdentifyEvent
import com.amplitude.core.events.BaseEvent
import com.amplitude.core.platform.DestinationPlugin
import com.segment.analytics.Analytics
import com.segment.analytics.Properties
import com.segment.analytics.Traits

/**
 * Segment destination plugin
 * @param appContext the Context
 * @param writeKey the segmentWriteKey defined in secrets.xml
 */
class SegmentDestinationPlugin(appContext: Context, writeKey: String) : DestinationPlugin() {
    var analytics: Analytics? = null
    val context: Context = appContext
    init {
        analytics = Analytics.Builder(appContext, writeKey).build()
    }

    override fun track(payload: BaseEvent): BaseEvent {
        val eventProperties =  Properties()
        payload.eventProperties?.forEach { entry -> entry.value?.let {
            eventProperties.put(entry.key,
                it
            )
        } }

        analytics?.track(payload.eventType, eventProperties)
        return payload
    }

    override fun identify(payload: IdentifyEvent): IdentifyEvent {
        val traits = Traits()
        payload.userProperties?.forEach { entry -> entry.value?.let {
            traits.put(entry.key,
                it
            )
        } }

        Analytics.with(context).identify(traits)
        return payload
    }

    override fun groupIdentify(payload: GroupIdentifyEvent): GroupIdentifyEvent {
        val traits = Traits()
        val groupId: String? = payload.groups?.keys?.first()

        payload.groupProperties?.forEach { entry -> entry.value?.let {
            traits.put(entry.key,
                it
            )
        } }

        if (groupId != null) {
            Analytics.with(context).group(groupId, traits)
        }

        return payload
    }
}